webanno/webanno
# #1755 - Support definition of constraints across slot features

**What's in the PR**
- Added support for constraint conditions to work across multi-valued features with other FSes as targets (i.e. slot features)
- Cleaned up code
- Upgraded JavaCC Maven plugin
- Added AnnotationBuilder and FeatureStructureBuilder helper classes in support module (quite useful for testing at least)

**How to test manually**
* Define a new span layer `EventTrigger` with
  * a slot feature called `participants` that targets the **Named entity** layer
  * a string feature called `value` which has a tagset with the tags `move` and `buy`
* Upload a document with e.g. text `Oracle bought Sun.`
* Annotate **Oracle** and **Sun** as **Named entities**, both with the value **ORG**
* Annotate **bought** as an **EventTrigger** and fill the named entities into its participant slots (role does not matter)
* In the project settings, import the following constraints:
```
import webanno.custom.EventTrigger as TRIGGER;

TRIGGER {
  participants.target.value = "PER" & participants.target.value = "LOC" -> value = "move";
  participants.target.value = "PER" & participants.target.value = "ORG" -> value = "buy";
  participants.target.value = "ORG" & participants.target.value = "ORG" -> value = "buy";
  participants.target.value = "PER" & participants.target.value = "PER" -> value = "buy";
}
```
* Go back to the annotation page, select the **EventTrigger** annotation and check that the **value** feature dropdown lists **buy** in bold at the top.

**Automatic testing**

**Documentation**
